### DISCUSSION TITLE
Experiencing "random" no response from modbus sensor.

### DISCUSSION THREAD

**Discussion by TristanLowell** [pymodbus-dev/pymodbus#2446] [Q&A] [answered]
**Issue**:
You can see from the logs that i am able to successfully read the sensor (modbus RTU RS485). However, at random times there seems to be "no response" and then when i try to decode the payload i get a modbus exception (that doesnt get caught as far as i can tell). I have messed with timeout and "sleep" time (polling interval) and have not found a combination that resolves the random communication issue. I think its the sensor may be in its own "polling interval" to get the next batch of data and a request is being sent to it at the same time? But i have not been able to confirm this is the issue. Here is the sensor documentation (https://info.bannerengineering.com/cs/groups/public/documents/literature/210732.pdf). 

I have provided code But it is not just a Modbus implementation, i am also using paho mqtt to send my modbus data to a remote broker. 

I am able to use modbus slave simulator and do not experience this issue, which leads me to believe it is he sensor that is the issue. I am reading the modbus register data from a json file. 

**Code**:
[mqtt_currImpl.txt](https://github.com/user-attachments/files/17669319/mqtt_currImpl.txt)
[pymod_currImpl.txt](https://github.com/user-attachments/files/17669320/pymod_currImpl.txt)
[main_currImpl.txt](https://github.com/user-attachments/files/17669321/main_currImpl.txt)

**Specs**:
OS: linux subsystem on Advantech Iot Cell router

pymodbus version. 3.7.4

**logs**:
DEBUG:paho.mqtt.client:Sending PUBLISH (d0, q1, r0, m23), 'b'device/a44058a7-e58f-438d-9578-71576b88cffb/nb/json'',                                                            properties=None, ... (945 bytes)
DEBUG:paho.mqtt.client:Received PUBACK (Mid: 23)
True
DEBUG:paho.mqtt.client:Sending PINGREQ
DEBUG:paho.mqtt.client:Received PINGRESP
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x0 0x31 0x0 0x1 0xd5 0xc5
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020219.353574
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0x0 0x0 0xb8 0x44
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0x0 0x0 0xb8 0x44
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[0]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x63 0x0 0x1 0x77 0x88
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020219.422162
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0x1 0xb0 0xb8 0x60
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0x1 0xb0 0xb8 0x60
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[432]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x64 0x0 0x1 0xc6 0x49
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020219.498806
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0x2 0x62 0x38 0xcd
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0x2 0x62 0x38 0xcd
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[610]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x65 0x0 0x1 0x97 0x89
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020219.570289
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0x0 0xc 0xb8 0x41
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0x0 0xc 0xb8 0x41
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[12]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x66 0x0 0x1 0x67 0x89
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020219.644268
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0x0 0x16 0x39 0x8a
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0x0 0x16 0x39 0x8a
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[22]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x67 0x0 0x1 0x36 0x49
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020219.722548
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0xd 0xc0 0xbc 0x84
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0xd 0xc0 0xbc 0x84
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[3520]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x68 0x0 0x1 0x6 0x4a
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020219.794497
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0xb 0x3 0xff 0x75
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0xb 0x3 0xff 0x75
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[2819]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x69 0x0 0x1 0x57 0x8a
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020219.867503
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0x0 0x6 0x38 0x46
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0x0 0x6 0x38 0x46
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[6]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x6a 0x0 0x1 0xa7 0x8a
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020219.941407
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0xfc 0x1b 0xb9 0x4f
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0xfc 0x1b 0xb9 0x4f
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[64539]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x95 0x0 0x1 0x97 0xba
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020220.01327
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0x2 0x20 0xb8 0xfc
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0x2 0x20 0xb8 0xfc
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[544]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x96 0x0 0x1 0x67 0xba
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020220.085011
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0x2 0x62 0x38 0xcd
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0x2 0x62 0x38 0xcd
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[610]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x97 0x0 0x1 0x36 0x7a
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020220.156744
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0x0 0x10 0xb9 0x88
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0x0 0x10 0xb9 0x88
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[16]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x98 0x0 0x1 0x6 0x79
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020220.232507
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0x0 0x1a 0x39 0x8f
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0x0 0x1a 0x39 0x8f
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[26]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x99 0x0 0x1 0x57 0xb9
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020220.304279
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0x11 0x2f 0xf5 0xc8
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0x11 0x2f 0xf5 0xc8
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[4399]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x9a 0x0 0x1 0xa7 0xb9
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020220.377571
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0xc 0x62 0x3c 0xad
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0xc 0x62 0x3c 0xad
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[3170]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x9b 0x0 0x1 0xf6 0x79
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020220.448801
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0x0 0x6 0x38 0x46
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0x0 0x6 0x38 0x46
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[6]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x9c 0x0 0x1 0x47 0xb8
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020220.520628
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0xff 0xec 0xf8 0x39
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0xff 0xec 0xf8 0x39
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[65516]
DEBUG:paho.mqtt.client:Sending PUBLISH (d0, q1, r0, m24), 'b'device/a44058a7-e58f-438d-9578-71576b88cffb/nb/json'',                                                            properties=None, ... (943 bytes)
DEBUG:paho.mqtt.client:Received PUBACK (Mid: 24)
True
DEBUG:paho.mqtt.client:Sending PINGREQ
DEBUG:paho.mqtt.client:Received PINGRESP
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x0 0x31 0x0 0x1 0xd5 0xc5
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020280.813568
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0x0 0x0 0xb8 0x44
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0x0 0x0 0xb8 0x44
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[0]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x63 0x0 0x1 0x77 0x88
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020280.885249
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x1 0x3 0x2 0x1 0xa2 0x38 0x6d
DEBUG:pymodbus.logging:Processing: 0x1 0x3 0x2 0x1 0xa2 0x38 0x6d
DEBUG:pymodbus.logging:decode PDU for 3
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 0
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
DEBUG:pymodbus.logging:[418]
get and verify data
DEBUG:pymodbus.logging:Current transaction state - TRANSACTION_COMPLETE
DEBUG:pymodbus.logging:Running transaction 0
DEBUG:pymodbus.logging:SEND: 0x1 0x3 0x9 0x64 0x0 0x1 0xc6 0x49
DEBUG:pymodbus.logging:Changing state to IDLE - Last Frame End - None Current Time stamp - 1731020280.957372
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Transaction failed. (Modbus Error: [Invalid Message] No response received, expected at least                                                            4 bytes (0 received))
DEBUG:pymodbus.logging:Processing:
DEBUG:pymodbus.logging:Getting transaction 0
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
Traceback (most recent call last):
  File "/root/mqttMbus/main.py", line 80, in <module>
    main()
  File "/root/mqttMbus/main.py", line 71, in main
    mbusDataPacket = mbus.mbusPollilng(mbus, slaves)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/mqttMbus/mbus/mbus.py", line 242, in mbusPollilng
    measureDict["data"] = {"n-value" : mbusObj.mbusConnect(SlaveDefFile[i]["Type"], SlaveDefFile[i]["Address"], Slav                                                           eDefFile[i]["NumRegisters"], SlaveDefFile[i]["SlaveId"], SlaveDefFile[i]["DataType"])}
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/mqttMbus/mbus/mbus.py", line 182, in mbusConnect
    decoder = self.initPayloadDecoder(rr, fc, self.endiness)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/mqttMbus/mbus/mbus.py", line 112, in initPayloadDecoder
    decoder = BinaryPayloadDecoder.fromRegisters(mbusRes.registers, byteorder=Endian.LITTLE, wordorder=Endian.BIG)
                                                 ^^^^^^^^^^^^^^^^^
AttributeError: 'ModbusIOException' object has no attribute 'registers'

**Comment by janiversen** [comment]
Your code is large, and not something I look into. If you have a problem then make some simple sample code that shows the problem.

I am pretty sure you do not get an exception thrown but a object returned that inherits from ModbusException. You should use the isError() function after the request call to verify you got a correct response.

Depending on your setup it is quite normal from time to time to get a bad or no response from your device.

**Comment by janiversen** [comment]
This exception "AttributeError: 'ModbusIOException' object has no attribute 'registers'" is because you address .registers without having verified the response is ok.

**Reply by TristanLowell** [reply]
Thank you for the help, I am wondering if there is a way to manually clear Tx or RX buffer to ensure extra characters are not being left over. I will get this error every so often: "Received Modbus library error(Modbus Error: [Input/Output] No Response received from the remote slave/Unable to decode response)". 

If this happens to a register it may not happen on the next polling interval. wondering if this is a buffer issue.

**Reply by janiversen** [reply]
You can see that with a debug log, pymodbus automatically skips garbage data in front of a sound frame.

**Reply by TristanLowell** [reply]
ok, thank you. just trying to get to the bottom of the issue. when a register address is passed to a read operation what format is the data expected to be in? can i just pass an integer representation or do i have to pass the value in HEX. Example if i want to read holding register at address 49 can i pass 49 or do i have to pass 0x31 (or just 31)?

**Reply by janiversen** [reply]
Look at the api signature it tells what is expected.

**Reply by TristanLowell** [reply]
Thank you, however the documentation is poor but i determined you can provide and integer value 34, 23, etc or you can provide a hex representation 0x31, 0x14, etc. I was wondering if an integer is passed does it get changed to a hex value? if a hex value is passed does it get converted at all or just kept as it?

I am able to get data from a specific address using modbus poll for my Banner QM30VT2 vibration temperature sensor, but when i use pymodbu i get a value of 0 back. I am using this code. reading only a single register (address 49).

import pymodbus.client as ModbusClient
from pymodbus.constants import Endian
from pymodbus.payload import BinaryPayloadDecoder
from pymodbus import (
    ExceptionResponse,
    FramerType,
    ModbusException,
    pymodbus_apply_logging_config
)

client = ModbusClient.ModbusSerialClient("COM3", FramerType.RTU, 9600, 8, "N", 1, timeout=5)
client.connect()
rr = client.read_holding_registers(48, 1, slave=1)
print(rr)
decoder = BinaryPayloadDecoder.fromRegisters(rr.registers, byteorder=Endian.LITTLE, wordorder=Endian.BIG)
out = decoder.decode_16bit_int()
print(out)

**Reply by TristanLowell** [reply]
ok, i have notice that the issue seems to be with how i am decoding. I have added a couple print statements so i can see the payload. i get this:

from pymodbus.constants import Endian
from pymodbus.payload import BinaryPayloadDecoder
from pymodbus import (
    ExceptionResponse,
    FramerType,
    ModbusException,
    pymodbus_apply_logging_config
)

client = ModbusClient.ModbusSerialClient("COM3", FramerType.RTU, 9600, 8, "N", 1, timeout=5)
client.connect()
rr = client.read_holding_registers(48, 1, slave=1)
print(rr) # added
print(rr.registers) # added
decoder = BinaryPayloadDecoder.fromRegisters(rr.registers, byteorder=Endian.LITTLE, wordorder=Endian.BIG)
out = decoder.decode_16bit_uint()
print(out)



ReadHoldingRegistersResponse (1)
[7833] <-- this is the proper value that i get when i use modbus poll
39198 <-- this is the value that i get when i decode. 

what am i doing wrong?

**Reply by janiversen** [reply]
What do you mean poor? the api signature identifies the possibilities uniquely!

Apart from that pull requests are welcome.

Your use of big/little endian seems wrong (but depends on your cpu). however why use the complicated payloader when we offer from/to int/float methods.

**Reply by TristanLowell** [reply]
I have a hard time understanding the documentation. some others may find it easy to follow. its not a put down to you im just saying for me its poor. My apologies if my understanding of how it is meant to be understood is lacking. im just tyring to solve my own issue.

what are the the use of bit and little endian was the only combination i could find to communicate with the RTU i am working with. What is "wrong" with my usage of them? how were they meant to be used?

what are the methods of "to/from" int/float? i do not think i have seen these options.

**Reply by janiversen** [reply]
What is difficult or poor in the signature read_holding(address: int, ...
I really do not see how that can be better, it states clearly only integers are accepted. If you are unsure if you can specify an int as 0x2345 is not a library problem but standard python knowledge.


I do not know which documentation you refer to, our online documentation contains the signatures alongside examples of how to use the API.

It's strange that you have a cpu with 2 endian encodings, normally it's either one or the other. Apart from that the registers contain integers in  the format used by your cpu.

If you have read the documentation you should have seen all the options.

**Reply by janiversen** [reply, marked as answer]
And fyi, saying the documentation is poor, is a problem for the project, if you do not understand it, there can be several reasons.

But instead of just calling it poor, submit a pull request to make it better....this is open source where you are expected to participate.
